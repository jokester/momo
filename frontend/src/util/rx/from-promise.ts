import { Observable, from, NEVER, concat } from 'rxjs';
import { concatAll, concatMapTo, dematerialize, materialize, switchMapTo, take } from 'rxjs/operators';

export function fromPromiseWithoutEnd<T>(p: PromiseLike<T>): Observable<T> {
  return concat(from(p).pipe(materialize(), take(1), dematerialize()), NEVER);
}
