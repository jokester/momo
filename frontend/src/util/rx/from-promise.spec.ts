import { from, NEVER, of, concat } from 'rxjs';
import { flatMap, materialize, switchMap, take, toArray } from 'rxjs/operators';
import { fromPromiseWithoutEnd } from './from-promise';

describe('from-promise', () => {
  it('rxjs', async () => {
    const fromRxjs1 = from(Promise.resolve(1)).pipe(materialize(), take(1), toArray());

    expect(await fromRxjs1.toPromise()).toMatchSnapshot('fromRxJs1');
  });

  it /* always timeout */.skip('transforms Promise to a 0|1', async () => {
    const p1 = fromPromiseWithoutEnd(Promise.resolve(1)).pipe(toArray());
    expect(await p1.toPromise()).toEqual([1, 2, 3]);
  });

  it('wtf1', async () => {
    expect(await from(Promise.resolve(1)).pipe(toArray()).toPromise()).toEqual([1]);
  });
});
