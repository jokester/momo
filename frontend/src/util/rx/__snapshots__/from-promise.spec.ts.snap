// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`from-promise rxjs: fromRxJs1 1`] = `
Array [
  Notification {
    "error": undefined,
    "hasValue": true,
    "kind": "N",
    "value": 1,
  },
]
`;
